File: part-2/week-1/depth_first_search/Graph.cc
#include <queue>
#include <stack>

#include "Graph.hh"

Graph::Graph(int n)
{
    this->n        = n;
    this->vertices = new Vertex[n];
}

Graph::~Graph(void)
{
    delete[] this->vertices;
}

void Graph::add_edge(int v, int w)
{
    // Adjust for indexing.
    v--;
    w--;

    // Add undirected edge.
    this->vertices[v].edges.push_front(w);
    this->vertices[w].edges.push_front(v);
}

void Graph::print(void)
{
    for (int i = 0; i < n; i++) {
        printf("%i -> ", i + 1);
        for (int j : this->vertices[i].edges)
            printf("%i ", j + 1);
        printf("\n");
    }
}

list<int> Graph::bfs(int s)
{
    // Create a queue to hold each vertex.
    queue<int> vertices;

    // List to hold all vertices.
    list<int> traversed_vertices;

    // Mark the current vertex as visited and add it to the queue.
    this->vertices[s - 1].is_visited = true;
    vertices.push(s - 1);

    // Iterate over all the vertices in the queue.
    while (!vertices.empty()) {
        // Get the first vertex from the queue.
        int v = vertices.front();
        vertices.pop();
        traversed_vertices.push_back(v + 1);

        // Add each unvisited vertex w that v is connected to, to the queue.
        for (int w : this->vertices[v].edges)
            if (!this->vertices[w].is_visited) {
                this->vertices[w].is_visited = true;
                vertices.push(w);
            }
    }

    // Mark all vertices as unvisited again.
    for (int i : traversed_vertices)
        this->vertices[i - 1].is_visited = false;

    return traversed_vertices;
}

list<int> Graph::dfs(int s)
{
    // Create a stack to hold each vertex.
    stack<int> vertices;

    // List to hold all vertices.
    list<int> traversed_vertices;

    // Mark the current vertex as visited and add it to the stack.
    this->vertices[s - 1].is_visited = true;
    vertices.push(s - 1);

    // Iterate over all the vertices in the stack.
    while (!vertices.empty()) {
        // Get the top vertex from the stack.
        int v = vertices.top();
        vertices.pop();
        traversed_vertices.push_back(v + 1);

        // Add each unvisited vertex w that v is connected to, to the stack.
        for (int w : this->vertices[v].edges)
            if (!this->vertices[w].is_visited) {
                this->vertices[w].is_visited = true;
                vertices.push(w);
            }
    }

    // Mark all vertices as unvisited again.
    for (int i : traversed_vertices)
        this->vertices[i - 1].is_visited = false;

    return traversed_vertices;
}

File: part-3/week-2/KruskalsMSTAlgorithm/Graph.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

class Graph {
  private class SortByEdgeLengths implements Comparator<MSTNode> {
    public int compare(MSTNode a, MSTNode b) {
      return a.getLength() - b.getLength();
    }
  }

  // Members.
  private int n;
  private Vertex[] vertices;

  Graph(int n) {
    this.n = n;
    this.vertices = new Vertex[n];
    for (int i = 0; i < n; i++) this.vertices[i] = new Vertex();
  }

  void addEdge(int tail, int head, int weight) {
    // Adjust for indexing.
    tail--;
    head--;

    // Add undirected edge between the vertices.
    this.vertices[tail].addEdge(head, weight);
    this.vertices[head].addEdge(tail, weight);
  }

  private ArrayList<MSTNode> getSortedListOfEdges() {
    // Create a list that will contain all edges (this list will contain duplicate edges too but it
    // won't matter).
    ArrayList<MSTNode> edges = new ArrayList<MSTNode>();

    // Add each edge to this list.
    for (int i = 0; i < this.n; i++)
      for (Edge e : this.vertices[i].edges)
        edges.add(new MSTNode(i, e.getToVertex(), e.getLength()));

    // Sort the given list in the increasing order of edge lengths.
    Collections.sort(edges, new SortByEdgeLengths());

    return edges;
  }

  ArrayList<MSTNode> getMinimumSpanningTree() {
    // Create an list containing all edges and their vertices.
    ArrayList<MSTNode> edges = this.getSortedListOfEdges();

    // Initialize a UnionFind structure.
    UnionFind uf = new UnionFind(this.n);

    // A list to hold the final MST.
    ArrayList<MSTNode> minimumSpanningTree = new ArrayList<MSTNode>();

    // Iterate through each of the sorted edges.
    for (MSTNode edge : edges) {
      // If the two vertices of this edge are not in the same subset, then add this edge to the MST.
      if (uf.find(edge.getFromVertex()) != uf.find(edge.getToVertex())) {
        uf.union(edge.getFromVertex(), edge.getToVertex());

        // Re-adjust for indexing.
        edge.setFromVertex(edge.getFromVertex() + 1);
        edge.setToVertex(edge.getToVertex() + 1);

        minimumSpanningTree.add(edge);
      }
    }

    return minimumSpanningTree;
  }

  /** Helper function for debugging. */
  void print() {
    for (int i = 0; i < n; i++) {
      System.out.print((i + 1) + " -> ");
      for (Edge e : this.vertices[i].edges)
        System.out.print((e.getToVertex() + 1) + "," + e.getLength() + " ");
      System.out.println();
    }
  }
}
